### PR TITLE
Update compiler plugin validation

### PR DESCRIPTION
**Pull request by kalaiyarasiganeshalingam**
## Purpose
$Subject

Resolves: [#1227](https://github.com/ballerina-platform/ballerina-standard-library/issues/1227)
## Checklist
- [x] Linked to an issue
- [x] Updated the changelog
- [x] Added tests

### CODE REVIEW

File: file-ballerina/CompilerPlugin.toml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+[plugin]
+id = "file-compiler-plugin"
+class = "io.ballerina.stdlib.file.compiler.FileCompilerPlugin"
+
+[[dependency]]
+path = "../file-compiler-plugin/build/libs/file-compiler-plugin-0.7.0-alpha8-SNAPSHOT.jar"
```
Comment (daneshk) [inline file-ballerina/CompilerPlugin.toml:6]: Shall we parameterized the version number?

File: file-compiler-plugin/src/main/java/io/ballerina/stdlib/file/compiler/FileServiceValidator.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    private static final String INVALID_REMOTE_FUNCTION = "Missing remote key word in the remote function `{0}`";
+    private static final String INVALID_FUNCTION_NAME = "Invalid function name `{0}`";
+    private static final String INVALID_RETURN_TYPE = "Return types are not allowed in the remote function `{0}`";
+    private static final String INVALID_PARAM_SIZE = "The remote function should contain a parameter";
+    private static final String EMPTY_SERVICE = "At least a single remote function required in the service";
+
+    @Override
+    public void perform(SyntaxNodeAnalysisContext syntaxNodeAnalysisContext) {
+        ServiceDeclarationNode serviceDeclarationNode = (ServiceDeclarationNode) syntaxNodeAnalysisContext.node();
+        long size = serviceDeclarationNode.members().stream().filter(child -> child.kind() ==
```
Comment (BuddhiWathsala) [inline file-compiler-plugin/src/main/java/io/ballerina/stdlib/file/compiler/FileServiceValidator.java:57]: Here we need to filter out only the file package-related services. 

You can refer to this commit https://github.com/ballerina-platform/module-ballerina-grpc/pull/158/commits/fecfeb3bf4fbfd10a95fa8ddc7442962ed172c01

**Review by BuddhiWathsala — CHANGES_REQUESTED** [review]

File: file-compiler-plugin/src/main/java/io/ballerina/stdlib/file/compiler/FileServiceValidator.java (hunk 2 above)
Comment (BuddhiWathsala) [reply file-compiler-plugin/src/main/java/io/ballerina/stdlib/file/compiler/FileServiceValidator.java:57]: Better if we can implement a test case to capture this use case, I mean a test case with multiple services including file listener.

File: file-compiler-plugin/src/main/java/io/ballerina/stdlib/file/compiler/FileServiceValidator.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+ */
+public class FileServiceValidator implements AnalysisTask<SyntaxNodeAnalysisContext> {
+    private static final String CODE = "File";
+    private static final String FILE_EVENT = "file:FileEvent";
+    private static final String RESOURCE_NAME_ON_CREATE = "onCreate";
+    private static final String RESOURCE_NAME_ON_DELETE = "onDelete";
+    private static final String RESOURCE_NAME_ON_MODIFY = "onModify";
+    private static final String INVALID_INPUT_PARAM = "Invalid parameter type `{0}` provided for remote function. " +
+            "Only file:FileEvent is allowed as the parameter type";
+    private static final String INVALID_REMOTE_FUNCTION = "Missing remote key word in the remote function `{0}`";
```
Comment (daneshk) [inline file-compiler-plugin/src/main/java/io/ballerina/stdlib/file/compiler/FileServiceValidator.java:48]:
```suggestion
    private static final String INVALID_REMOTE_FUNCTION = "Missing remote keyword in the remote function `{0}`";
```

File: file-compiler-plugin/src/main/java/io/ballerina/stdlib/file/compiler/FileServiceValidator.java (hunk 2 above)
Comment (kalaiyarasiganeshalingam) [reply file-compiler-plugin/src/main/java/io/ballerina/stdlib/file/compiler/FileServiceValidator.java]: Fixed in 4d82650

File: changelog.md [hunk 4]
```diff
@@ -7,6 +7,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ## [1.1.0-alpha8] - 2021-04-09
 
 ### Changed
+- [Update the compiler plugin to validate file service](https://github.com/ballerina-platform/ballerina-standard-library/issues/1227)
 - [[Change the path field in the `ListenerConfig` from optional to mandatory]](https://github.com/ballerina-platform/ballerina-standard-library/issues/1235)
```
Comment (kalaiyarasiganeshalingam) [inline changelog.md:11]:
```suggestion
- [Change the path field in the `ListenerConfig` from optional to mandatory](https://github.com/ballerina-platform/ballerina-standard-library/issues/1235)
```

**Review by BuddhiWathsala — APPROVED** [review]
LGTM

File: file-compiler-plugin-test/src/test/java/io/ballerina/stdlib/file/compiler/CompilerPluginTest.java [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+        Assert.assertEquals(diagnosticResult.diagnostics().size(), 0);
+    }
+
+    @Test
+    public void testCompilerPluginWithInvalidParams() {
+        Package currentPackage = loadPackage("package_02");
+        PackageCompilation compilation = currentPackage.getCompilation();
+        String errMsg = "ERROR [file_service.bal:(29:4,31:5)] The remote function should contain a parameter";
+        String errMsg1 = "ERROR [file_service.bal:(33:4,35:5)] Invalid function name `onEdit`";
+        String errMsg2 = "ERROR [file_service.bal:(37:4,39:5)] The remote function should contain a parameter";
```
Comment (daneshk) [inline file-compiler-plugin-test/src/test/java/io/ballerina/stdlib/file/compiler/CompilerPluginTest.java:59]: Shall we mention which parameter is missing here. Something like `The remote function should only contain `file:FileEvent ` parameter`

File: file-compiler-plugin-test/src/test/java/io/ballerina/stdlib/file/compiler/CompilerPluginTest.java [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+        DiagnosticResult diagnosticResult = compilation.diagnosticResult();
+        Assert.assertEquals(diagnosticResult.diagnostics().size(), 0);
+    }
+
+    @Test
+    public void testCompilerPluginWithInvalidParams() {
+        Package currentPackage = loadPackage("package_02");
+        PackageCompilation compilation = currentPackage.getCompilation();
+        String errMsg = "ERROR [file_service.bal:(29:4,31:5)] The remote function should contain a parameter";
+        String errMsg1 = "ERROR [file_service.bal:(33:4,35:5)] Invalid function name `onEdit`";
```
Comment (daneshk) [inline file-compiler-plugin-test/src/test/java/io/ballerina/stdlib/file/compiler/CompilerPluginTest.java:58]: Shall we improve the message. Something like `Invalid function name `onEdit`, File Listener only supports `onCreate`, `onModify` and `onDelete` remote functions`

File: file-compiler-plugin-test/src/test/java/io/ballerina/stdlib/file/compiler/CompilerPluginTest.java (hunk 5 above)
Comment (daneshk) [reply file-compiler-plugin-test/src/test/java/io/ballerina/stdlib/file/compiler/CompilerPluginTest.java:59]: Shall we make all error messages start in lowercase align with compiler error messages which is lower cases

File: file-compiler-plugin-test/src/test/java/io/ballerina/stdlib/file/compiler/CompilerPluginTest.java [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+    }
+
+    @Test
+    public void testCompilerPluginWithInvalidParams() {
+        Package currentPackage = loadPackage("package_02");
+        PackageCompilation compilation = currentPackage.getCompilation();
+        String errMsg = "ERROR [file_service.bal:(29:4,31:5)] the remote function should only contain " +
+                "file:FileEvent parameter";
+        String errMsg1 = "ERROR [file_service.bal:(33:4,35:5)] invalid function name `onEdit`, " +
+                "File Listener only supports `onCreate`, `onModify` and `onDelete` remote functions";
```
Comment (daneshk) [inline file-compiler-plugin-test/src/test/java/io/ballerina/stdlib/file/compiler/CompilerPluginTest.java:60]:
```suggestion
                "file listener only supports `onCreate`, `onModify` and `onDelete` remote functions";
```

**Review by daneshk — APPROVED** [review]
LGTM

**Pull request merged**